brush-core/src/builtins/ulimit.rs

```diff
@@ -0,0 +1,206 @@
+        short: 'c',
+        unit: Unit::Block,
+    };
+    const FSIZE: ResourceDescription = ResourceDescription {
+        resource: Resource::FSIZE,
+        description: "file size",
+        short: 'f',
+        unit: Unit::Block,
+    };
+    const RSS: ResourceDescription = ResourceDescription {
+        resource: Resource::RSS,
+        description: "max memory size",
+        short: 'm',
+        unit: Unit::KBytes,
+    };
+    const NOFILE: ResourceDescription = ResourceDescription {
+        resource: Resource::NOFILE,
+        description: "open files",
+        short: 'n',
+        unit: Unit::Number,
+    };
+
+    fn get(&self, hard: bool) -> std::io::Result<String> {
+        let val = if hard {
+            self.resource.get_hard()?
+        } else {
+            self.resource.get_soft()?
+        };
+
+        if val == rlimit::INFINITY {
+            Ok("unlimited".into())
+        } else {
+            Ok(format!("{val}"))
+        }
+    }
+
+    fn set(&self, set_hard: bool, value: LimitValue) -> std::io::Result<()> {
+        let (soft, hard) = self.resource.get()?;
+        let value = match value {
+            LimitValue::Soft => soft,
+            LimitValue::Hard => hard,
+            LimitValue::Unlimited => rlimit::INFINITY,
+            LimitValue::Value(v) => v,
+            LimitValue::Unset => return Ok(()),
+        };
+
+        if set_hard {
+            self.resource.set(soft, value)
+        } else {
+            self.resource.set(value, hard)
+        }
+    }
+
+    /// Print either soft or hard limit
+    fn print(&self, hard: bool) -> std::io::Result<()> {
+        let unit = match self.unit {
+            Unit::Block => format!("(block, -{})", self.short),
+            Unit::KBytes => format!("(kbytes, -{})", self.short),
+            Unit::Number => format!("(-{})", self.short),
+        };
+        let resource = self.get(hard)?;
+        println!("{:<26}{:>16} {}", self.description, unit, resource);
+        Ok(())
+    }
+}
+
+const ALL_RESOURCES: [ResourceDescription; 4] = [
+    ResourceDescription::CORE,
+    ResourceDescription::FSIZE,
+    ResourceDescription::NOFILE,
+    ResourceDescription::RSS,
+];
+
+#[derive(Debug, Clone, Copy)]
+enum LimitValue {
+    Unset,
+    Unlimited,
+    Soft,
+    Hard,
+    Value(u64),
+}
+
+impl FromStr for LimitValue {
+    type Err = <u64 as FromStr>::Err;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let v = match s {
+            "" => LimitValue::Unset,
+            "unlimited" => LimitValue::Unlimited,
+            "soft" => LimitValue::Soft,
+            "hard" => LimitValue::Hard,
+            _ => LimitValue::Value(s.parse()?),
+        };
+        Ok(v)
+    }
+}
+
+/// Modify shell resource limits.
+///
+/// Provides control over the resources available to the shell and processes
+/// it creates, on systems that allow such control.
+#[derive(Parser, Debug)]
+pub(crate) struct ULimitCommand {
```

Comment:
For most built-ins we've implemented, we've tried to leave placeholder stubs for all of the options -- so runtime use of them results in an "unimplemented error" that we can get reports on and track (vs. argument parsing errors), e.g.:

https://github.com/reubeno/brush/blob/7a7c642454aa0e26027d2abb436d3c7ba40dd8fc/brush-core/src/builtins/mapfile.rs#L57-L59

I'm supportive of getting this change in as incremental progress without implementing the other options but would like to see similar "unimplemented stubs" here. Would that be feasible?

Reply:
let me add one

Reply:
Overall takes more or less the same time to complete it though...

Reply:
Got it. I agree that, if it takes just as much time to add the stub as the real implementation, then we'd appreciate the real implementation :smile:

_(I wasn't sure if you were intentionally limiting the implementation because (a) you wanted to leave some work for other interested contributors, (b) you were short on time to look at the other options, or (c) the other options were trickier. All of those are, of course, reasonable explanations.)_

Reply:
Mostly `a`, plus running out of time that weekend. All the options have potentially only the annoyance of having to pick the right name for the resource since some are also synonyms, but it literally boils down to fill in the structure and add the line to use it.

Reply:
I definitely appreciate you being supportive of other contributors :smile:.

I'd prefer stubs of some kind, but in this case I'd compromise on at least getting placeholder `// TODO` comment(s) so we have some way to track the missing functionality and hold ourselves accountable for getting it filled in.

Reply:
Once I have another timeslice I'll fill in more, shouldn't take much either way. We might need some creative solution to deal with synonyms or contribute the failover to the rlimit crate itself.